prow/secretutil/censor.go

```diff
@@ -0,0 +1,103 @@
+/*
+Copyright 2021 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// Package secretutil implements utilities to operate on secret data.
+package secretutil
+
+import (
+	"encoding/base64"
+	"strings"
+	"sync"
+
+	"go4.org/bytereplacer"
+)
+
+// Censorer knows how to replace sensitive data from input.
+type Censorer interface {
+	// Censor will remove sensitive data previously registered with the Censorer
+	// from the input. This is thread-safe, will mutate the input and will never
+	// change the overall size of the input.
+	Censor(input []byte)
+}
+
+func NewCensorer() *ReloadingCensorer {
+	return &ReloadingCensorer{
+		RWMutex:  &sync.RWMutex{},
+		Replacer: bytereplacer.New(),
+	}
+}
+
+type ReloadingCensorer struct {
+	*sync.RWMutex
+	*bytereplacer.Replacer
+}
+
+var _ Censorer = &ReloadingCensorer{}
+
+// Censor will remove sensitive data previously registered with the Censorer
+// from the input. This is thread-safe, will mutate the input and will never
+// change the overall size of the input.
+func (c *ReloadingCensorer) Censor(input []byte) {
+	c.RLock()
+	// we know our replacer will never have to allocate, as our replacements
+	// are the same size as what they're replacing, so we can throw away
+	// the return value from Replace()
+	c.Replacer.Replace(input)
+	c.RUnlock()
+}
+
+// RefreshBytes refreshes the set of secrets that we censor.
+func (c *ReloadingCensorer) RefreshBytes(secrets ...[]byte) {
+	var asStrings []string
+	for _, secret := range secrets {
+		asStrings = append(asStrings, string(secret))
+	}
+	c.Refresh(asStrings...)
+}
+
+// Refresh refreshes the set of secrets that we censor.
+func (c *ReloadingCensorer) Refresh(secrets ...string) {
+	var replacements []string
+	addReplacement := func(s string) {
+		replacements = append(replacements, s, strings.Repeat(`*`, len(s)))
```

Comment:
Instead of always telling the length of the original secret, can we make this the `min(len(s), 3)` to avoid exposing that info in the 99% of cases where the secret is longer than three characters?

Reply:
In the context of the streaming censoring algorithm for artifacts, allowing this to mutate the length of the output is not ideal. A quick overview of the algorithm again:

1. read in 2N bytes where N is the longest secret to censor
2. censor all secrets in those 2N bytes
3. write the first N bytes to the output file
4. move the last N bytes to the start of the buffer
5. read N more bytes into the buffer
6. loop from 2

So if censoring changes the length, things get math-y and hard there. I agree that your suggestion is slightly nicer - do you think it's worth the extra complexity it will incur later?

Reply:
Leaking length is technically still a leak b/c it reduces the searching space significantly. I think the complexity is worth it.

Reply:
Leaking length limits the search space significantly so I think the added complexity would be worth it.

Reply:
The library can't mutate the length of the slice itself, so it returns e.g.:

```
secret SECRET c2VjcmV0 sEcReT
*** SECRET *** sEcReT0 sEcReT
<--- replacement ---><suffix>
```

I'd need to post-process the slice walking backwards to trim it. This seems like needless complexity but I'll give it a shot.

Reply:
@petr-muller how does it matter that an attacker sees that a token is 30 chars long for e.g. a GitHub OAuth token? I don't buy the "search space" argument - not only would they need to, after seeing an artifact with a censored bit, determine *what* it's censoring, what system it is for, then they have ... no information?

Like, all GitHub OAuth tokens are of the same length. Knowing that our token is also that length gives them info enough to search .. the entire space. While in a purely technical sense sure it leaks something, I just don't see the utility of it.

Reply:
And, actually, post-processing the slice to trim the suffix requires us to re-allocate the memory anyway, since we need to keep the old slice around for the traversal. I'm leaning hard against doing this.

Reply:
After more communication, it's clear that it will not be possible to achieve this without changing the byte replacer itself. That's orders of magnitude harder than this pull. I've left the signature here with `*[]byte` to allow it to be done in the future, but it won't be part of this initial pull.